airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py

```diff
@@ -115,56 +140,58 @@ def test_discover(capsys, tmp_path, json_spec, scenario):
         assert expected_msg in get_error_message_from_exc(exc)
     else:
-        output = read(capsys, tmp_path, scenario)
-        records, logs = output["records"], output["logs"]
-        assert len(records) == len(expected_records)
-        assert len(logs) == len(expected_logs)
-        assert_expected_records_match_output(records, expected_records)
-        assert_expected_logs_match_output(logs, expected_logs)
-
-
-def assert_expected_records_match_output(output: List[Dict[str, Any]], expected_output: List[Dict[str, Any]]):
-    for actual, expected in zip(output, expected_output):
-        assert actual["record"]["data"] == expected["data"]
-        assert actual["record"]["stream"] == expected["stream"]
+        output = read(capsys, caplog, tmp_path, scenario)
+        _verify_read_output(output, scenario)
 
 
-def assert_expected_logs_match_output(logs: List[Dict[str, Any]], expected_logs: List[Dict[str, Any]]):
-    for actual, expected in zip(logs, expected_logs):
-        assert actual["log"]["level"] == expected["level"]
-        assert actual["log"]["message"] == expected["message"]
-
-
-def run_test_read_incremental(capsys, tmp_path, scenario):
+def run_test_read_incremental(capsys, caplog, tmp_path, scenario):
     expected_exc, expected_msg = scenario.expected_read_error
     if expected_exc:
         with pytest.raises(expected_exc):
-            read_with_state(capsys, tmp_path, scenario)
+            read_with_state(capsys, caplog, tmp_path, scenario)
     else:
-        output = read_with_state(capsys, tmp_path, scenario)
-        expected_output = scenario.expected_records
-        assert len(output) == len(expected_output)
-        for actual, expected in zip(output, expected_output):
-            if "record" in actual:
-                assert actual["record"]["data"] == expected
-            elif "state" in actual:
-                assert actual["state"]["data"] == expected
+        output = read_with_state(capsys, caplog, tmp_path, scenario)
+        _verify_read_output(output, scenario)
+
+
+def _verify_read_output(output, scenario):
+    records, logs = output["records"], output["logs"]
+    logs = [log for log in logs if log.get("level") in ("ERROR", "WARN", "WARNING")]
+    expected_records = scenario.expected_records
+    assert len(records) == len(expected_records)
+    for actual, expected in zip(records, expected_records):
+        if "record" in actual:
+            assert actual["record"]["data"] == expected["data"]
+            assert actual["record"]["stream"] == expected["stream"]
+        elif "state" in actual:
+            assert actual["state"]["data"] == expected
+
+    if scenario.expected_logs:
+        expected_logs = scenario.expected_logs.get("read", [])
+        assert len(logs) == len(expected_logs)
+        _verify_expected_logs(logs, expected_logs)
+
+
+def _verify_expected_logs(logs: List[Dict[str, Any]], expected_logs: List[Dict[str, Any]]):
```

Comment:
I probably added a bit more validation of logs than we really need. I was using it in my own manual testing, but if it feels too brittle we can take it out.